### ISSUE TITLE
[breaking change] Modify the 'UP' algorithm to handle intersection types earlier

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#54002]
### Change Intent

Change the [`UP`](https://github.com/dart-lang/language/blob/main/resources/type-system/upper-lower-bounds.md#upper-bounds) algorithm which is used to compute the standard upper bound (also known as the least upper bound) of two types, such that it handles intersection types just before the rules about `Null`, rather than after.

### Justification

This change ensures that the algorithm never produces a result of the form `(X & B)?`. Note that according to the relevant specification document, intersection types like `X & B` ..

> can never appear as sub-components of other types,

as described [near the end of this section](https://github.com/dart-lang/language/blob/main/resources/type-system/subtyping.md#types). In particular, this implies that they can never appear in a type of the form `(X & B)?`. This modification of `UP` ensures that no such malformed types are created.

Importantly, the existence of such types have created some issues, e.g., the soundness issue in https://github.com/dart-lang/sdk/issues/53887.

As an aside, here is another reason why we would benefit from this change:

Type inference using the new rules will use the promoted bound in the upper bound of two types where one is an intersection type. This means that the resulting expression can have a larger interface (that is, we have a larger number of members that we can invoke on it), and this is very much in line with a typical motivation for promoting a variable.

```dart
bool condition = ...;

void f<X extends num>(X x) {
  if (x is int) { // `x` now has type `X & int`.
    var y = condition ? x : null;
    y?.isEven;
  }
}
```

With the current behavior, `y` gets the inferred type `(X & int)?`, which is a problematic type in the first place. [One proposal](dart-lang/language#2427) to avoid the malformed type would make it `X?`, and [the proposal which is the subject of this issue](dart-lang/language#3435) makes it `int?`. This makes `y?.isEven` correct based on the proposal of this issue whereas it would be a compile-time error with the other proposal.

### Impact

Changes to `UP` must always be considered carefully, because they will change the behavior of static analysis in many situations, and hence they may cause widespread breakage.

However, this particular change has been seen to create very little breakage. The change is implemented in https://dart-review.googlesource.com/c/sdk/+/333922. This CL changes the implementation of the CFE and the analyzer to get the proposed behavior. It changes two tests (which are directed at this particular behavior, so they must be updated). Finally, it changes 3 SDK libraries where a rather unusual expression is used to detect the JavaScript value `undefined` and replace it by `null`.

A global run on internal code (TAP presubmit) came up with one failure, requiring changes to 4 lines in a single function body, apart from failures in the tests mentioned above.

In other words, presumably, the modified `UP` breaks the code only in very few cases.

### Mitigation

Here is the only case that came up in practice (the same function had 4 lines that are very similar, we just show one of them):

```dart
T? nullifyZeroOrEmpty<T>(T? value) {
  if (value is int) return value == 0 ? null : value;
  ...
}

// Mitigation
T? nullifyZeroOrEmpty<T>(T? value) {
  if (value is int) return value == 0 ? null as T? : value;
  ...
}
```

### Change Timeline

This change can be performed when needed, the code in https://dart-review.googlesource.com/c/sdk/+/333922 is ready for review. It would be beneficial if it can be done before the end of November in order to have the feature in place for a while before the stable release.

### Associated CLs

Spec change: https://github.com/dart-lang/language/pull/3435.
Implementation: https://dart-review.googlesource.com/c/sdk/+/333922.

**Comment by itsjustkevin** [comment]
@grouma @vsmenon @Hixie could you take a look at this breaking change request?

**Comment by Hixie** [comment]
It would be useful to have some more examples of things this improves. The `nullifyZeroOrEmpty` mitigation seems like a step backwards, but with more context showing the steps forwards, I suspect I would find it insignificant? Hard for me to tell from the proposal.

**Comment by Hixie** [comment]
(I don't really understand the `nullifyZeroOrEmpty` case really, but I'm guessing it's because `... ? ... : ...` has weird behaviour unrelated to this and this change just triggers that weird behaviour?)

**Comment by grouma** [comment]
> A global run on internal code (TAP presubmit) came up with one failure, requiring changes to 4 lines in a single function body, apart from failures in the tests mentioned above.

I assume these changes were landed?

This request SGTM assuming we repeat the TGP prior to landing.

**Comment by eernstg** [comment]
@Hixie wrote:
> The `nullifyZeroOrEmpty` mitigation seems like a step backwards,

First, I think it's important to note that the current behavior introduces types where an intersection type is a subterm of another type (in particular, `(X & B)?` for some type variable `X` which was promoted to `B`, and then the type as a whole was made nullable by `UP`). They are just malformed types. In that sense it's a problem at the same level as allowing `typedef MyType = 42;`, and we should simply ensure that our type operations do not produce types for which we haven't defined what they might mean, or which rules we should apply to them.

We did have some discussions about the feasibility of extending the meta-theory such that this kind of type could be added to the type universe, but that would be a non-trivial exercise. If anyone thinks this would be a really useful generalization then we should discuss it as a separate issue, and as a new feature, and we should spend the required time and effort to explore the consequences and make sure that we maintain soundness etc. I think it's fair to say that this is not an option right now, at this time we just need to make sure we don't have malformed types in our static analysis.

Note that the current treatment does give rise to soundness issues in the analyzer, as reported in https://github.com/dart-lang/sdk/issues/53887.

With that in mind, we should repair the part of `UP` that creates types of the form `(X & B)?`.

The change which is being scrutinized here does this by moving the rules about intersection types up in the definition of `UP`, such that they occur before the rules about `Null` and the rules about nullable types (the rules about intersection types currently occur after all these null-related rules).

In other words, the _only situation where this update causes anything to change_ is a situation where `UP` is computed on an intersection type and a type which is either (1) `Null` (plus a few other syntactic forms which are also the type `Null`, e.g., `Null?` or `Never?`) or (2) a type of the form `T?` (so it's some, but not all, nullable types).

As mentioned, I found only 4 lines with an error after introducing the change which is being scrutinized here (TAP presubmit), so it is really a rarely visited corner of the language. I've only seen breakage in a situation where a conditional expression has `null` as one of the branches (like `b ? e1 : null`).

However, it _is_ possible to find an example where it makes a difference without using null:

```dart
class A<X> {}

var condition = true;

void f<X extends A<X>?>(X x, A<X>? ax) {
  if (x is A<X>) { // `x` now has type `X & A<X>`.
    var y = condition ? x : ax; // Type of `y` shown below.
  }
}

void main() {}

/*
Using the old rules, type of `y`:
  UP(X & A<X>, A<X>?) =
  UP(X & A<X>, A<X>)? =
  UP(A<Object?>, A<X>)? =
  A<Object?>?.

Using the rules under scrutiny here, type of `y`:
  UP(X & A<X>, A<X>?) =
  A<X>?
*/
```

In summary:
- We need to correct `UP` such that it does not yield malformed types.
- The change under scrutiny here has a very small amount of breakage; only `b ? e1 : null` has been seen in practice.
- We can construct examples where `null` doesn't occur (but `?` types do), where the change does makes a difference. This requires a very special configuration of types with and without `?`. The typing improves in the given example, but it's not easy to quantify whether it's going to be typical. The main point is that this is very unlikely to occur with any significant frequency in practice.

**Comment by eernstg** [comment]
@grouma wrote:
> I assume these changes were landed?

Not yet, but they will be landed before the changes to the analyzer and CFE are landed.

**Comment by eernstg** [comment]
@a-siva, I guess you added the label 'area-specification' because there's a need for _some_ 'area-...' label to indicate that this issue has been triaged?

The breaking change as such is about a change to the language, that is, a change to the specification _and_ every tool that uses the `UP` algorithm. This makes me think that 'area-specification' is somewhat misleading, and 'area-meta' is more in line with the actual purpose of this issue.

So I'll change the label to 'area-meta'. If this creates a problem for you then please revert that change (and please write a comment in this thread in order to give us a heads-up about why it must be 'area-specification' ;-).

**Comment by leafpetersen** [comment]
Ping on this.  This is an inconsistency between our tools (and a soundness issue in at least one of them, fortunately just the analyzer I believe) so I'd like to get it resolved soon.  Are there objections to this change?  If so can we discuss them on #53887, along with other possible resolutions?

**Comment by Hixie** [comment]
I have no objection, but I must admit to not really understanding the implications. I'd love to see more concrete examples of differences (especially improvements) that this causes.

**Comment by eernstg** [comment]
@grouma wrote:
> I assume these changes were landed?

The preparations for landing the CL that updates the `UP` algorithm are available here:

- https://dart-review.googlesource.com/c/sdk/+/337726, updating a language test that currently succeeds because it has wrong expectations about the static type of some expressions, and the current implementation happens to expect exactly that (this change is justified because of the existing as well as the updated definition of `UP`, but the change that implements the new definition of `UP` happens to fix this bug as well). So we should land this CL _even_ if this breaking change request is rejected (but in that case we need to fix the tools in some other way).
- https://dart-review.googlesource.com/c/sdk/+/337780, updating 3 internal hash map libraries, because they contain, in total, 6 expressions that are type correct under the old definition of `UP` and type incorrect under the new definition. However, all 6 expressions are type correct with both the old and the new definition of `UP`, which means that this change is benign even in the case where this breaking change request is rejected.

In other words, we can be prepared for this change, and the preparations do not introduce any issues in both cases: No issues if we accept the breaking change, and only issues that are real bugs if we reject the breaking change.

**Comment by eernstg** [comment]
> I'd love to see more concrete examples of differences (especially improvements) that this causes.

First, I'd include "don't produce types that aren't well-formed" as an improvement.

In order to find situations where this change of `UP` causes software properties to change (e.g., from "compiles and runs" to "has compile-time error"), we need to consider situations where the reordering of rules matters. (That's the only thing we do, nothing changes other than the ordering of some rules in the definition of `UP`.)

So we need to consider situations that currently match one of the following cases, but in the future will match the intersection type cases. (I've eliminated rules about legacy types because they do not exist any longer.)

<details><summary>Rules that used to occur before intersection, but will now occur after.</summary>

- **UP**(`T1`, `T2`) where **NULL**(`T1`) and **NULL**(`T2`) =
  - `T2` if **MOREBOTTOM**(`T1`, `T2`)
  - `T1` otherwise

- **UP**(`T1`, `T2`) where **NULL**(`T1`) =
  - `T2` if  `T2` is nullable
  - `T2?` otherwise

- **UP**(`T1`, `T2`) where **NULL**(`T2`) =
  - `T1` if  `T1` is nullable
  - `T1?` otherwise

- **UP**(`T1`, `T2`) where **OBJECT**(`T1`) and **OBJECT**(`T2`) =
  - `T1` if **MORETOP**(`T1`, `T2`)
  - `T2` otherwise

- **UP**(`T1`, `T2`) where **OBJECT**(`T1`) =
  - `T1` if `T2` is non-nullable
  - `T1?` otherwise

- **UP**(`T1`, `T2`) where **OBJECT**(`T2`) =
  - `T2` if `T1` is non-nullable
  - `T2?` otherwise

- **UP**(`T1?`, `T2?`) = `S?` where `S` is **UP**(`T1`, `T2`)
- **UP**(`T1?`, `T2`) = `S?` where `S` is **UP**(`T1`, `T2`)
- **UP**(`T1`, `T2?`) = `S?` where `S` is **UP**(`T1`, `T2`)
</details>

For well-formed types, an intersection type can neither be a **NULL** type nor an **OBJECT** type nor a type of the form `T?`. This implies that only the following rules can partake in this scenario (where the applicability of each rule can be restricted by extra conditions, i.e., it's the same rule but we know more about the situation where it can occur):

<details><summary>Rules before intersection, now after, and not impossible.</summary>

- **UP**(`T1`, `T2`) where **NULL**(`T1`) and `T2` is `X2 & B2` =
  - `T2` if  `T2` is nullable
  - `T2?` otherwise

- **UP**(`T1`, `T2`) where **NULL**(`T2`) and `T1` is `X1 & B1` =
  - `T1` if  `T1` is nullable
  - `T1?` otherwise

- **UP**(`T1`, `T2`) where **OBJECT**(`T1`) and `T2` is `X2 & B2` =
  - `T1` if `T2` is non-nullable
  - `T1?` otherwise

- **UP**(`T1`, `T2`) where **OBJECT**(`T2`) and `T1` is `X1 & B1` =
  - `T2` if `T1` is non-nullable
  - `T2?` otherwise

- **UP**(`T1?`, `T2`) where `T2` is `X2 & B2` = `S?` where `S` is **UP**(`T1`, `T2`)
- **UP**(`T1`, `T2?`) where `T1` is `X1 & B1` = `S?` where `S` is **UP**(`T1`, `T2`)
</details>

We can simplify these rules a bit, because we know that a type of the form `X & B` cannot be nullable, and it is non-nullable if and only if `B` is non-nullable:

<details><summary>Relevant and possible rules, simplified.</summary>

- **UP**(`T1`, `T2`) where **NULL**(`T1`) and `T2` is `X2 & B2` = `(X2 & B2)?`.
- **UP**(`T1`, `T2`) where **NULL**(`T2`) and `T1` is `X1 & B1` = `(X1 & B1)?`.

- **UP**(`T1`, `T2`) where **OBJECT**(`T1`) and `T2` is `X2 & B2` =
  - `Object` if `B2` is non-nullable
  - `Object?` otherwise

- **UP**(`T1`, `T2`) where **OBJECT**(`T2`) and `T1` is `X1 & B1` =
  - `Object` if `B1` is non-nullable
  - `Object?` otherwise

- **UP**(`T1?`, `T2`) where `T2` is `X2 & B2` = `S?` where `S` is **UP**(`T1`, `T2`)
- **UP**(`T1`, `T2?`) where `T1` is `X1 & B1` = `S?` where `S` is **UP**(`T1`, `T2`)
</details>

The first two rules produce a type which is not well-formed, so it's inherently an improvement that the new definition of `UP` will not have those cases (because we will use an intersection case when `T2` is `X2 & B2` respectively `T1` is `X1 & B1`).

For the next two rules we can change `T1`/`T2` to `Object`, because **OBJECT**(`T1`) respectively **OBJECT**(`T2`) implies that the given type is both a supertype and a subtype of `Object`. In other words, it _is_ the type `Object`, possibly spelled differently. 

Finally, the rules are symmetric (both the intersection type rules and these rules), so we only have to consider one of the two directions. This yields yet another simplified set of rules.

Relevant and possible rules, simplified twice:

- **UP**(`X1 & B1`, `T2`) where **OBJECT**(`T2`) =
  - `Object` if `B1` is non-nullable.
  - `Object?` otherwise.

- **UP**(`X1 & B1`, `T2?`) = `S?` where `S` is **UP**(`X1 & B1`, `T2`).

So the only situations where this change matters are the ones where the code uses one of these two rules, and in the future it will instead use the intersection type rule:

- **UP**(`X1 & B1`, `T2`) =
  - `T2` if `X1 <: T2`
  - otherwise `X1` if `T2 <: X1`
  - otherwise **UP**(`B1a`, `T2`)
    where `B1a` is the greatest closure of `B1` with respect to `X1`,
    as defined in [inference.md].

We need a small lemma in order to proceed: If `T <: Object` and `X` is a type variable, and `T` is not `X`, and `S` is the greatest closure of `T`, then `S <: Object`. This can be checked by a sketching a proof by induction about the rules [here](https://github.com/dart-lang/language/blob/main/resources/type-system/inference.md#type-variable-elimination-least-and-greatest-closure-of-a-type).

*So we're lucky that `X extends X` is a compile-time error. ;-)*

Now we can see how this proposal can change the outcome of a computation of `UP`:

(1) Consider the situation where the **OBJECT** rule is used today, that is `T2 == Object` .

(1.1) In the case where `B1` is non-nullable, we used to get the result `Object`. With the proposed change we will use the intersection rule. We know that the bound `B` of `X1` must be a supertype of `B1`, but this doesn't constrain `B` to be nullable or non-nullable. Assume that `B` is non-nullable; then `X1 <: B <: Object == T2`, and the result is still `Object`. Otherwise, `B` is potentially nullable and we do _not_ have `X1 <: T2`, and we certainly don't have `T2 <: X1` either, so we get the result `UP(B1a, Object)`, which must be `Object` or a top type. However, we know that `B1 <: Object`, so the lemma proves that `B1a <: Object`, and hence the result is still `Object`.

(1.2) In the case where `B1` is potentially nullable (that's the negation of '`B1` is non-nullable'), we used to get `Object?`. We know that `B1 <: T2` does _not_ hold, and we certainly don't have `T2 <: X1` either, so we get the result `UP(B1a, Object)` again, which must still be `Object` or a top type. This implies that the result is the same as before the change, or it used to be `Object?` and it is now `Object` (an improvement).

(2) Consider the situation where the `T2?` rule is used with the old rules. We used to get the result `UP(X1 & B1, T2)?`.

By induction, `UP(X1 & B1, T2)` must either be preserved, or it has changed because it matches one of the other cases that we're considering here (we know that `T2` isn't a type of the form `T3?` itself, so we can't use that rule again). But in those other cases we either get the same result or an improvement from `Object?` to `Object`; but the result is then `Object` and then we re-add the `?`, so the final result in this case is `Object?`, that is, no change.

Modulo mistakes in the reasoning, this implies that the new definition of `UP` preserves the results we have today in all cases where the old rules yield a result which is not malformed.

Let's reconsider the example that I found a few days ago:

```dart
T? nullifyZeroOrEmpty<T>(T? value) {
  if (value is int) return value == 0 ? null : value;
  ...
}
```

Here is an expanded version of this example, showing that it's an exercise in juggling malformed types:

```dart
var b = true;

T? nullifyZeroOrEmpty<T>(T? value) {
  if (value is int) {
    // `value` promoted to `T` then to `T & int`:
    T v1 = value;
    int v2 = value;
    // `UP(Null, T & int) == (T & int)?`, malformed, so:
    // `v3` could have type `T?` promoted to `(T & int)?`, maybe?
    var v3 = value == 0 ? null : value;
    v3.isEven; // No analyzer error, contradicts type `(T & int)?`.
    return v3; // OK, also if return type is `int?`.
  }
  ...
}
```

So if we agree that we shouldn't have these malformed types then I actually don't see any regressions. We only have one known example, and it basically shows that it may be convenient to use the malformed type, but that's not a very promising situation to be in, because we can't trust the rules about the malformed types.

**Comment by leafpetersen** [comment]
> I have no objection, but I must admit to not really understanding the implications. I'd love to see more concrete examples of differences (especially improvements) that this causes.

@Hixie I think this is the wrong way of thinking about this change - the *direct* user facing implications are expected to be quite minimal either way.  The point of the change is that the existing specified algorithm for upper bounds is buggy - it essentially causes undefined behavior by introducing a form of type which the rest of the meta-theory assumes as an invariant will never exist (and hence for which there is no defined behavior).  Rather than try to extend the meta-theory to account for more general intersection types, which is potentially a quite significant undertaking with unknown implications, we are instead proposing simply to fix the upper bound algorithm to not produce the ill-formed type.  This is likely to be almost entirely invisible to users - the cases where the ill-formed types are produced (and then consumed in various unspecified and hence implementation defined ways) are quite rare (empirically, based on testing the change on internal code).  The kinds of places that you would expect to see a change are where:
  - You have a local variable (or in general any promotable variable), let's call it `x`.
  -  The type of `x` is a type variable `X`.
  - You do an instance check on `x` against a type `T`, which when it succeeds proves that `x` has both type `X`, and type `T`
     - Note that we model this by promoting `x` to a a restricted intersection type `X & T`.
   - You perform some operation which causes an upper bound to be computed for the type of `x` ('X & T`) and some other nullable type (e.g. `Null`).  So for example, `var a = (true) ? x : null` induces such an upper bound computation.

Prior to this change, the type produced in examples like the above will have the form `(X & T)?`.  After this change, we will eliminate the intersection, producing `T?`.  In general, the intersection type **is a more useful type**.  It preserves strictly more information.  But general intersection types (like general union types) are: quite complicated at the meta-theory level; algorithmically expensive; and very hard to give good error messages for.  And so we have chosen (so far) to restrict our uses of them to very controlled places: we don't allow them to appear as nested types, but only as the top level types of variables, and even then only in very restricted cases.   We've considered tackling general union types + intersection types as a language feature in the past, but for the time being would prefer to maintain our existing limited use of them.  This change restores the meta-theory to a well-understood state given that intended level of support.

**Comment by eernstg** [comment]
https://dart-review.googlesource.com/c/sdk/+/337780 has now landed.

**Comment by Hixie** [comment]
If this has no impact on users then it's probably not a breaking change and I have no objection. :-)

**Comment by eernstg** [comment]
Very good! @itsjustkevin, do you need further input in order to approve this?

**Comment by itsjustkevin** [comment]
Approving this.  @vsmenon please chime in if you object!

**Comment by eernstg** [comment]
https://dart-review.googlesource.com/c/sdk/+/337726 has now been landed. This is the CL that changes a language test to expect the correct static type of an expression in 4 locations; this needed to be done independently of this breaking change, because those 4 expressions will have the corrected static types with this breaking change as well as without this breaking change. (It's just that the current implementation of `UP` erroneously succeeded with the types that we expected before this CL.)

The updates that introduce the actual breaking change (the modified implementation of `UP` in the analyzer and in the CFE) have _not_ yet been landed; they await possible objections.

**Comment by vsmenon** [comment]
lgtm

**Comment by kevmoo** [comment]
FYI: I hit this with the following code.

```dart
T? emptyAsNull<T>(T value) {
  if (value is Iterable) {
    return (value.isEmpty) ? null : value;
  } else if (value is Map) {
    return (value.isEmpty) ? null : value;
  }
  return value;
}
```

The workaround wasn't bad, but it was *weird* to hit this is what I thought was VERY stable code

```dart
T? emptyAsNull<T>(T value) {
  if (value is Iterable) {
    return (value.isEmpty) ? null as T? : value;
  } else if (value is Map) {
    return (value.isEmpty) ? null as T? : value;
  }
  return value;
}
```

@natebosch took it to 11

```dart
T? emptyAsNull<T>(T value) => switch (value) {
      Iterable(isEmpty: true) || Map(isEmpty: true) => null,
      _ => value,
    };
```

**Comment by eernstg** [comment]
At this time we still only have failures in one very specific kind of code:

An expression of the form `b ? e1 : e2` where `e1` or `e2` is `null` has a more general type after the breaking change, and that creates a type error in the enclosing construct. (For instance `T v = b ? e1 : e2;` may now be an error because the type of the initializing expression is no longer assignable to the declared type `T`. Or `var v = b ? e1 : e2;` may create errors where `v` is used, because it gets an inferred type which is more general than before this breaking change).

In other words, all known failures created by this breaking change can be handled as described by @kevmoo [here](https://github.com/dart-lang/sdk/issues/54002#issuecomment-1852577095).

If you encounter a different kind of failure due to this breaking change then it would be greatly appreciated if you write a comment here about what it is, and how you fixed it!

**Comment by leafpetersen** [comment]
> If you encounter a different kind of failure due to this breaking change then it would be greatly appreciated if you write a comment here about what it is, and how you fixed it!

To be clear, we're interested in hearing from folks who encounter this in the above form as well, since that will give us early signal if this turns out to be more impactful than we expect.

**Comment by eernstg** [comment]
Closing: https://dart-review.googlesource.com/c/sdk/+/333922 has landed. Added the 'implementation' label to this issue because there was no separate implementation issue.
